### DISCUSSION TITLE
apiflask and nginx

### DISCUSSION THREAD

**Discussion by wangleiofficial** [apiflask/apiflask#307] [Q&A] [answered]
The location of openapi is not loaded correctly when using nginx as port proxy
```
location /XXXXXX/ {
       proxy_pass http://localhost:7777/;
    }
```
![image](https://user-images.githubusercontent.com/58798173/171352706-a229438c-fa70-446e-b38b-e96f7d739b76.png)

but debug mode works fine.
![image](https://user-images.githubusercontent.com/58798173/171352636-f0cd7a3b-7b9b-4679-9470-cc8d741f92af.png)

**Comment by wangleiofficial** [comment]
openapi ignores routes set by nginx, which leads to an error

**Comment by wangleiofficial** [comment, marked as answer]
When running behind a reverse proxy, not only is the public URL different, but often you want the application to appear below some path other than /. Sometimes you even use https from the outside, in which case you likely don't use SSL between the proxy server and your Flask app.

In such cases, if you can control the headers in the proxy server (easy with Nginx and many others), this middleware lets you transparently change where the application appears. This works dynamically for each request, so you can simultaneously access from the local machine using http://localhost:port without using the full URL.
```
class ReverseProxied(object):
    '''Wrap the application in this middleware and configure the 
    front-end server to add these headers, to let you quietly bind 
    this to a URL other than / and to an HTTP scheme that is 
    different than what is used locally.

    In nginx:
    location /myprefix {
        proxy_pass http://192.168.0.1:5001;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Scheme $scheme;
        proxy_set_header X-Script-Name /myprefix;
        }

    :param app: the WSGI application
    '''
    def __init__(self, app):
        self.app = app

    def __call__(self, environ, start_response):
        script_name = environ.get('HTTP_X_SCRIPT_NAME', '')
        if script_name:
            environ['SCRIPT_NAME'] = script_name
            path_info = environ['PATH_INFO']
            if path_info.startswith(script_name):
                environ['PATH_INFO'] = path_info[len(script_name):]

        scheme = environ.get('HTTP_X_SCHEME', '')
        if scheme:
            environ['wsgi.url_scheme'] = scheme
        return self.app(environ, start_response)
```

Install in app using:
```

    app = Flask(__name__)
    app.wsgi_app = ReverseProxied(app.wsgi_app)
```

**Comment by vba34520** [comment]
This method is capable of displaying whe API documentation correctly. It is very clever, Thanks!

But there is a fly in the ointment that it can't use the Curl on the API documentation, so do [my method](https://github.com/apiflask/apiflask/issues/377).

**Comment by greyli** [comment]
I think the `ReverseProxied` can be replaced with `werkzeug.middleware.proxy_fix.ProxyFix`:

```py
from werkzeug.middleware.proxy_fix import ProxyFix
from apiflask import APIFlask


app = APIFlask(__name__)
app.wsgi_app = ProxyFix(
    app.wsgi_app, x_for=1, x_proto=1, x_host=1, x_prefix=1
)
```

Then set a proper servers field:

```py
app.servers = [{'url': '/myprefix'}]
```

See the details here: https://apiflask.com/tips/#running-the-application-behide-a-reverse-proxy

P.S. With APIFlask 1.2.1, the `servers` field will be set automatically (#395).
